src/mscorlib/shared/System/IO/Path.Unix.cs

```diff
@@ -69,6 +69,35 @@ private static string RemoveLongPathPrefix(string path)
             return path; // nop.  There's nothing special about "long" paths on Unix.
         }
 
+        private static void GetTempPath(ref ValueStringBuilder builder)
+        {
+            const string TempEnvVar = "TMPDIR";
+            const string DefaultTempPath = "/tmp/";
+
+            // Get the temp path from the TMPDIR environment variable.                        
+            int requiredSize = 0;
+            while ((requiredSize = Microsoft.Win32.Win32Native.GetEnvironmentVariable(TempEnvVar, builder.Span)) > builder.Capacity)
```

Comment:
@JeremyKuhne @Anipik  to avoid allocation of string i need to use `GetEnvironmentVariable(string lpName, Span<char> lpValue)`. I've exposed Span property and i'm pretty sure you don't want this. Can you help me with overload for `GetEnvironmentVariable(string lpName, ref char lpTempFileName, int size)` or `GetEnvironmentVariable(string lpName, ValueStringBuilder builder)` i don't find a way to cast `ref char` to `char*`

Reply:
You need to use `fixed` to get the address (`char*`) of a ref char. If you follow the code down you'll see:

``` C#
[DllImport(Interop.Libraries.Kernel32, CharSet = CharSet.Auto, SetLastError = true, BestFitMapping = false)]
private static extern unsafe int GetEnvironmentVariable(string lpName, char* lpValue, int size);

internal static unsafe int GetEnvironmentVariable(string lpName, Span<char> lpValue)
{
    fixed (char* lpValuePtr = &MemoryMarshal.GetReference(lpValue))
    {
        return GetEnvironmentVariable(lpName, lpValuePtr, lpValue.Length);
    }
}
```

What you should really do here is rewrite `GetEnvironmentVariableCoreHelper` to take a ValueStringBuilder.  Take a look at the changes I made in https://github.com/dotnet/corefx/commit/050bc33738887d9d8fcc9bc5965b7d9ca65bc7f4 for examples.

`char*` is equivalent to `ref char` in P\Invoke signatures. As such, the helper in Win32Native can just be removed and the Pinvoke can be changed to take a `ref char` as everything is calling it with a `Span`. You can avoid having to use the fixed statement that way. (Note that there is currently an issue in CoreRT that requires manually using fixed- if you look at P\Invokes in the shared folder you'll see `#ifdef`s that do it both ways.)

Reply:
Ok TYVM

Reply:
You will have issues with making the same code build with CoreRT. The environment stuff is factored differently there. We may need a (public) no-allocation API to read the environment to make this work well. No-allocation environment access is likely a lot more useful than no-allocation temporary path generator.

Reply:
> You will have issues with making the same code build with CoreRT.

Ah yes, I forgot that while the P/Invoke isn't shared, this is. We can do the same hack on the P/Invoke there that I mentioned above no? (e.g. ifdef and manually pin the ref char)

> We may need a (public) no-allocation API to read the environment to make this work well.

There isn't a need for it to be public to unblock this, is there? I agree that we should have a public API for it eventually.

Reply:
Correct, it is not strictly blocking. But it would be best to be introducing the public no-allocation flavors of APIs in priority order (most valuable/impactful APIs first).

Reply:
@JeremyKuhne  @jkotas when you talk about public non-allocation API to read environment, do you refer to add overload to Environment class or to a new api?

Reply:
Overload much like this one, i.e. `Environment.TryGetEnvironmentVariable`.  We'd still want an internal one that takes `ValueStringBuilder` that said API would call.

Reply:
We should be using the same patterns internally as we are using for out internal APIs. If the `Try*` pattern that returns bool and Span does not work for out internal use, it is unlikely going to work well for our users as well. This suggests that the pattern we are using for the public API is wrong.

This applies to both GetEnvironmentVariable or GetTempPath APIs.

I think we need to step back and do work on the API design. We may need to expose the ValueStringBuilder as public to make this work well.

Reply:
@JeremyKuhne i go on with above idea(paying attention to corert) right? If i success with this could help also on new environment api.

Reply:
Sorry @JeremyKuhne  i've got a delay on refresh messages, i read your comment now. I wait your decision to understand how go on!I stay here and "listen".

Reply:
@jkotas I'll open an issue on ValueStringBuilder and link to it. While I think we definitely need VSB type APIs, I'm not confident that not having the existing `Try*` and `Span` pattern becomes moot with said APIs. What if you want to write to fixed size span and don't want it grabbing from array pool? Does it make sense to have an option on ValueStringBuilder that doesn't allow it to grow? I'll discuss it in the issue.

Reply:
https://github.com/dotnet/corefx/issues/28379